### ISSUE TITLE
SC 2.4.11 and out-of-viewport content

### ISSUE THREAD

**Issue by pomerantsev** [w3c/wcag#2204]
The [Not fully obscured](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum.html#unobscured) section of 2.4.11 talks about overlapping elements.

Should it also include elements that get pushed out of view on, say, a `scroll` event?

For example, headers with fixed positioning that get hidden when user scrolls down seem to be quite common: https://codepen.io/pavelp/full/qBVbVKm

The issue in the above example is that, if tabbing starts when the page is scrolled down a little, the focused element will be out of view. It would be quite easy to fix (for example, by listening to the `focusin` event), but I don't believe this case is currently covered by any of the WCAG 2.2 SCs.

Happy to be wrong though!

**Comment by alastc** [comment]
I'm not sure if this is necessarily something under author control?

Generally when you tab to something and it gains focus, the browser will automatically scroll it into view. It doesn't necessarily scroll the entire focus indicator into view so the "not-completely" is useful for that.

Also, the SC is scoped to when things "receive" keyboard focus, so if it subsequently gets moved out of view, that isn't in scope. (If the user hits tab, that should bring the focus back into tab.)

I'm not convinced at the moment that encouraging developers to manually keep elements in view would be a good idea overall, it could lead to some odd, counterproductive effects.

**Comment by alastc** [comment]
Also, if you can tab to something completely out of view (e.g. hidden), that would fail focus-visible already, so the new SC doesn't need to try and cover that.

If that makes sense, are you happy to close the issue?

**Comment by pomerantsev** [comment]
Thanks, Alastair!

Before we close, I'd like to make sure we're on the same page.

> Generally when you tab to something and it gains focus, the browser will automatically scroll it into view.

> Also, the SC is scoped to when things "receive" keyboard focus, so if it subsequently gets moved out of view, that isn't in scope.

I agree with both of the above!

> Also, if you can tab to something completely out of view (e.g. hidden), that would fail focus-visible already, so the new SC doesn't need to try and cover that.

From the user's standpoint, how is being "completely out of view" different from "completely obscured by another element" (the latter being covered by 2.4.11)? The outcome seems to be the same: they can't see the focused element.

Also, my example doesn't seem to fail 2.4.7: the focus indicator is visible if the element is visible, and 2.4.7 talks specifically about the indicator.

> In the fixed-position demo, when selecting a menu item the focus should move to the appropriate section (or fail focus-order), so the question then becomes how do you tab back to the navigation?

Apologies if I'm not understanding your comment well enough, but the demo is not about what happens when a menu item gets activated. There's nothing there altering the focus order, and tabbing to navigation is never prevented.

What's happening (and what I'm still not fully convinced that other SCs guard against) is that when this fixed-position element with a CSS transform gets focus, it _may_ be out of view, and bringing it into view on focus without affecting focus order or the page's scroll position is as simple as the following code snippet:
```js
header.addEventListener('focusin', () => {
  header.classList.remove('hidden');
});
```
Here's the full example with the above fix: https://codepen.io/pavelp/full/oNozEOV

**Comment by JAWS-test** [comment]
@alastc 

> Also, if you can tab to something completely out of view (e.g. hidden), that would fail focus-visible already, so the new SC doesn't need to try and cover that.

If this is true, then we wouldn't need the additions to focus visibility at 2.4.11 and 2.4.12 at all, because it makes no difference whether the focus is not visible because the element is positioned outside the visible area or is inside the visible area but obscured. I.e. an invisible focus due to an invisible element should then always be a violation of 2.4.7.

**Comment by alastc** [comment]
Just coming back to this one whilst prepping the survey.

> my example doesn't seem to fail 2.4.7: the focus indicator is visible if the element is visible, and 2.4.7 talks specifically about the indicator.

But both are hidden, I'm not sure why that's not applicable as the indicator is not visible?


> how is being "completely out of view" different from "completely obscured by another element"

By completely out of view, I meant they are not within the viewport rather than hidden by other content. From a user's point of view, that doesn't matter. But it does mean something different from a developer point of view, it isn't other content blocking it (like a fixed footer), you need to specifically prevent the browser showing the focused items.


> If this is true, then we wouldn't need the additions to focus visibility at 2.4.11 and 2.4.12 at all,

2.3.11 defines minimum 'visibility' of the indicator, and 2.4.12 covers the scenario where the indicator is within the viewport but covered by other content. (We had a long discussion about that, and the group didn't think that scenario was covered by 2.4.7.)

**Comment by alastc** [comment]
Proposed response:

<hr>

The [example](https://codepen.io/pavelp/full/qBVbVKm) shows a scenario where the navigation is hidden when the page is scrolled down a bit, and shift-tabbing back up the page can progress through the navigation without it being visible.

As the links are not being hidden by other page content it is not covered by the 'unobscured' SC. However, if the indicators are not visible at all, and not covered by other content, that would fail 2.4.7.

**Comment by mbgower** [comment]
I'll weigh in a bit on this one. In discussion leading up to scoping WCAG 2.2, AGWG members on a call were not able to agree that something would fail [2.4.7 Focus Visible](https://www.w3.org/TR/WCAG21/#focus-visible) if the element receiving was not in the viewport BUT did have a focus indicator.
The language of the SC reads:
>  Any keyboard operable user interface has a mode of operation where the keyboard focus indicator is visible.

Some argued that in the case of a floating footer obscuring the item with focus, if the user could arrow up to bring the item with focus into view, then "a mode of operation" existed where the keyboard focus indicator is visible.

The combination of the "when receiving keyboard focus" and the "not obscured" language were designed to make it explicit that any author-created content, such as a floating footer, that keeps the item receiving focus from being visible in the viewport, would constitute a failure.

**Comment by mbgower** [comment]
A follow up. Maybe the word "By" is getting us in trouble on this wording:
> The item with focus is not entirely hidden by author-created content.

If it was "due to" I think that may make it clearer that the author causing something to be invisible for instance, would also fail. It's not just 'covering' it up. :)

**Comment by mbgower** [comment]
Proposed response:
Thank you for your comment. As [described in detail above](https://github.com/w3c/wcag/issues/2204#issuecomment-1085961926), while there was general AGWG member agreement that it _is_ an accessibility problem if the author has to scroll around to locate the component that has just received focus (i.e., if it does not automatically appear in the viewport), there was no consensus that the language of Focus Visible currently results in a failure of that situation.

The wording of the Focus Appearance SC you mentioned is specifically designed to make that situation result in a failure of WCAG.

In the time since your issue was opened, this wording has been broken out as a separate Success Criterion, in order to cut down on the complexity of Focus Appearance (Minimum) and make the requirements clearer and separate.

**Comment by pomerantsev** [comment]
@mbgower this is great news! I also think that breaking out visibility of the element itself into a separate success criterion is the right move. This is in fact what we've done internally at Squarespace because focus appearance and element visibility on focus seemed to be separate concerns.

**Comment by alastc** [comment]
The [response above](https://github.com/w3c/wcag/issues/2204#issuecomment-1096984939) was agreed by the group.
https://www.w3.org/2022/04/19-ag-minutes.html#item06
